src/main/java/org/spongepowered/api/item/properties/AbstractItemProperty.java

```diff
@@ -0,0 +1,177 @@
+     * @param value The value of the property
+     */
+    protected AbstractItemProperty(@Nullable V value) {
+        this(value, null);
+    }
+
+    /**
+     * Initialise the value to the specified value and use the specified
+     * operator, use the default key.
+     *
+     * @param value The property value
+     * @param op The operator for the property
+     */
+    protected AbstractItemProperty(@Nullable V value, Operator op) {
+        this(null, value, op);
+    }
+
+    /**
+     * Use the specified key and value and set operator to the default.
+     *
+     * @param key The key identifying the property
+     * @param value The property value
+     */
+    protected AbstractItemProperty(@Nullable K key, @Nullable V value) {
+        this(key, value, null);
+    }
+
+    protected AbstractItemProperty(@Nullable K key, @Nullable V value, Operator op) {
+        this.key = key != null ? key : this.getDefaultKey(value);
+        this.value = value;
+        this.operator = op != null ? op : this.getDefaultOperator(this.key, value);
+    }
+
+    /**
+     * Get the default value for {@link #key}, used in case null is passed in
+     * (since we can't have a null key). In general this should return the class
+     * name of the property itself but subclasses are free to alter this
+     * behaviour if they wish.
+     *
+     * @param value Value passed in to the ctor, supplied in case a subclass
+     *      wants to return a specific default key based on the value
+     * @return default key to use. Must not be null!
+     */
+    @SuppressWarnings("unchecked")
+    protected K getDefaultKey(@Nullable V value) {
+        return (K) this.getClass().getSimpleName();
+    }
+
+    /**
+     * Return the default operator to use, based on the supplied key and value.
+     *
+     * @param key Property key
+     * @param value Property initial value, may be null
+     * @return operator to use
+     */
+    protected Operator getDefaultOperator(K key, @Nullable V value) {
+        return Operator.defaultOperator();
+    }
+
+    @Override
+    public K getKey() {
+        return this.key;
+    }
+
+    @Override
+    public V getValue() {
+        return this.value;
+    }
+
+    @Override
+    public Operator getOperator() {
+        return this.operator;
+    }
+
+    @Override
+    public boolean matches(@Nullable ItemProperty<?, ?> other) {
+        return this.getOperator().compare(this, other);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean equals(Object obj) {
+        if (!(obj instanceof ItemProperty)) {
+            return false;
+        }
+
+        ItemProperty<?, ?> other = (ItemProperty<?, ?>) obj;
+        if (!other.getKey().equals(this.getKey())) {
+            return false;
+        }
+
+        return other.getValue().equals(this.getValue());
+    }
+
+    @Override
+    public int hashCode() {
+        return this.hashCodeOf(this.getKey()) ^ this.hashCodeOf(this.getValue()) * 37;
```

Comment:
Why do you use such hash code method with an additional utility method instead of just `Objects.hashCode` which also supports null values?